formbricks/formbricks
# feat: use redis for rate limiting & next caching to resolve memory issues

## What does this PR do?
- Resolves Memory Leak due to Middleware LRU Cache
- Simplifies Rate Limiting in Cloud by using a simpler int approach rather than an array
- Introduces the use of Redis for Internal Caching + Rate Limiting instead of in-heap storage for reducing memory consumption spike.
- Introduces a new env var `REDIS_CLIENT_URL` that users can use to set

## How should this be tested?
Check Memory Usage on running both when
1. Deployed on Vercel
2. Deployed on Self Hosted Instance with & without the Redis Env var.

It should improve significantly & the GC should also periodically bring it down in the load drops!

## Fixes the Memory Leak based on the below findings

1. Load tested the current self hosting infra with around 300 reqs/sec with 10 concurrent users and could notice a **SIGNIFICANT BUMP** in memory (reached from **200 MB to 1 GB** in a few minutes!)

https://github.com/formbricks/formbricks/assets/55556994/45eac4e5-3369-402f-a57b-eb355b251762

2. And on **dropping the load to 0 reqs/sec**, the Node Garbage Collector (GC) cleaned things up but **could NOT bring it under 600 MB**.
![Screenshot_2024-02-09-13-31-21_1920x1080](https://github.com/formbricks/formbricks/assets/55556994/2f03258b-3abb-4419-9b5b-53cc1afa821d)

3. We then commented the codebase of a couple of endpoints to understand where the memory leak is actually happening ie infra/code basically narrowing down the scope and our finding was:

```
Even on returning an empty {} json as response from sync endpoint and removed any logic/service calls happening! We still see the RAM usage gradually increasing with the same amount of load
```
![Screenshot_2024-02-09-13-59-12_1920x1080](https://github.com/formbricks/formbricks/assets/55556994/ab1613bf-3c12-4db9-8e8b-2adeefcfe0ed)

4. Now, from this, we inferred that the memory leak isnt in our core business logic.
5. On further looking for previous witnesses on memory leaks in nextjs on Github/StackOverFlow, many pointers led us to **Image Optimisation** & **Sharp** package dependency.
> So for that we tested and it and got the following conclusion: There's still gradual increase in memory even when the load is constant but the increase is comparatively lesser than when used w next/image (took 2x time to reach 1 GB of memory (16 minutes)). On stopping the load, the memory consumptions drops to around half ie 512 MB.

This also did not feel like our solution!

6. We now decided to go back to the basics and use our very own **Node's Debugger** to find out things internally at runtime. We started with the **JS Heap size** and found that as the **load is increased** and over time, the **total memory usage to heap usage ratio keeps on decreasing**, meaning the **memory keeps on growing**!

https://github.com/formbricks/formbricks/assets/55556994/2f9e6d37-56c5-4740-9cff-e692726db3db

7. Now on further debugging inside the Heap looking at the Deltas between the snapshots, we found something unusual:

![image](https://github.com/formbricks/formbricks/assets/55556994/2603deb8-5023-443f-a1f4-bf5f8f2d77f2)

**Our `middleware.js` had a ton of closures & array calls!**

8. So now going deeper into the rabbit hole, the next thing to find was why was this happening ie if these are per-requests, the GC should clean them up at a decent rate post response. So we now used the **`--trace-gc` for node** to see how the GC is performing. Turned out the **GC was working fine**.

9. Now we get back to debugging the **heap snapshots** and then found out that **as the number of different IP users increased, the heap size increased**, hinting at our **internal `LRU Cache`** that we used!

10. Next debug step was to **comment out the `middleware.js` logic** and see how the memory leak was performing and boooom, the **memory handled the load pretty well** without going crazy! We found our **CULPRIT**!

11. Now, after a lot of ideation and discussions internally, we implemented the `webdis` (Redis HTTP client + server) because the **Nextjs middleware is compiled to run on edge runtime** (serverless, cloudflare workers)(it cannot connect with Redis using its own protocol or on top of raw TCP) so now on self hosting instances, we recommend to use Redis for rate limiting buckets!

![image (5)](https://github.com/formbricks/formbricks/assets/55556994/fc7b3c3f-cd3e-4058-be9c-e79c4d01b597)

> With this, 
- The **RAM usage is stable** (increases gradually (way slower than previous times) but drops when GC runs in a few mins majorly)
- The Webdis (Redis server + client) **does not take more than 10 MB of RAM**
- Even after around 1-2 hours of such constant load of around **800 concurrent users and 150 reqs/sec**, the **memory did not shoot up**
![Screenshot_2024-02-14-21-49-30_1920x1080 (1)](https://github.com/formbricks/formbricks/assets/55556994/7761cf4a-b429-41f4-a752-5e053b7df347)
- And once the load is dropped, in a few minutes, the **GC cleans things up**!

![Screenshot_2024-02-14-21-51-04_1920x1080](https://github.com/formbricks/formbricks/assets/55556994/4724ab1f-a71f-419d-bc87-4b832b66e785)

All the end users needs to do is:
Run the below service in their existing docker-compose.yml:
```
  webdis:
    image: nicolas/webdis
    ports:
      - 7379:7379
    restart: unless-stopped
    <<: *environment
```

And pass this env var to Formbricks: `REDIS_HTTP_CLIENT_URL: http://webdis:7379`

## Checklist

### Required

### Appreciated
